main.js

```diff
@@ -1,129 +1,87 @@
+const { Cu } = require("chrome");
+const { devtools } =
+  Cu.import("resource://gre/modules/devtools/Loader.jsm", {});
 const devtoolsRequire = devtools.require;
-const {ConnectionManager} = devtoolsRequire("devtools/client/connection-manager");
-let {Devices} = Cu.import("resource://gre/modules/devtools/Devices.jsm");
-const {when: unload} = require("sdk/system/unload");
+const { defineLazyGetter } = devtoolsRequire("devtools/toolkit/DevToolsUtils");
+
+const events = require("sdk/event/core");
+const { when: unload } = require("sdk/system/unload");
+const { Devices } = Cu.import("resource://gre/modules/devtools/Devices.jsm");
+const { gDevToolsBrowser } =
+  Cu.import("resource:///modules/devtools/gDevTools.jsm");
+defineLazyGetter(this, "adb", () => {
+  return require("./adb");
+});
+defineLazyGetter(this, "fastboot", () => {
+  return require("./fastboot");
+});
+defineLazyGetter(this, "Device", () => {
+  return require("./device");
+});
 
-let promise;
-try {
-  promise = Cu.import("resource://gre/modules/commonjs/promise/core.js").Promise;
-} catch (e) {
-  promise = Cu.import("resource://gre/modules/commonjs/sdk/core/promise.js").Promise;
+// Set this right away on startup
+Devices.helperAddonInstalled = true;
+unload(() => Devices.helperAddonInstalled = false);
+
+// In Firefox 39+ startup can be deferred until WebIDE is opened
+if (gDevToolsBrowser.isWebIDEInitialized) {
+  gDevToolsBrowser.isWebIDEInitialized.promise.then(start);
+} else {
+  start();
 }
 
-// As of Firefox 36, WebIDE exposes an API to register new runtimes.
-try {
+// Main startup sequence to listen for new devices, inform WebIDE, etc.
+function start() {
+  // As of Firefox 36, WebIDE exposes an API to register new runtimes.
   let Runtimes = devtoolsRequire("devtools/webide/runtimes");
   if (Runtimes && Runtimes.RuntimeScanners) {
     let scanner = require("./scanner");
     scanner.register();
   }
-} catch (e) {}
-
-Devices.helperAddonInstalled = true;
-exports.shutdown = function() {
-  Devices.helperAddonInstalled = false;
-  adb.stop(true);
-};
 
-adb.start().then(function () {
-  adb.trackDevices();
-});
+  adb.start().then(function() {
+    adb.trackDevices();
+  });
+  unload(() => adb.stop(true));
 
-/**
- * A Device instance is created and registered with the Devices module whenever
- * ADB notices a new device is connected.
- *
- * Any changes here should be examined carefully for backwards compatibility
- * issues.  Other add-ons, like the Tools Adapter, make use of these low-level
- * device objects too.
- */
-function Device(id) {
-  this.id = id;
-}
+  function onFastbootStart() {
+    fastboot.startPolling();
+  }
+  Devices.on("fastboot-start-polling", onFastbootStart);
+  unload(() => Devices.off("fastboot-start-polling", onFastbootStart));
+  function onFastbootStop() {
+    fastboot.stopPolling();
+  }
+  Devices.on("fastboot-stop-polling", onFastbootStop);
+  unload(() => Devices.on("fastboot-stop-polling", onFastbootStop));
```

Comment:
I just realized. For now, the b2g addon isn't integrated into WebIDE, so that it is expected to work even if you don't use/open WebIDE.
We should move the fastboot listener out of `start`.